src/core/ddsc/tests/psmx.c

```diff
@@ -1264,6 +1326,378 @@ CU_Test (ddsc_psmx, basic)
+          }
+        }
+        {
+          PsmxType1* sample;
+          for (int i = 0; i < sample_cnt; ++i) {
+            result = dds_request_loan(writer2, (void**)&sample);
+            CU_ASSERT_FATAL(result == DDS_RETCODE_OK);
+            sample->xy.x = i;
+            sample->xy.y = (uint8_t)i + 1;
+            sample->z = (uint8_t)i + 2;
+            result = dds_write(writer2, sample);
+            CU_ASSERT_FATAL(result == DDS_RETCODE_OK);
+          }
+        }
+
+        {
+          SC_Model* samples1[sample_cnt];
+          dds_sample_info_t sinfo[sample_cnt];
+          const dds_entity_t ws1 = dds_create_waitset(participant);
+          dds_entity_t rcond1 = dds_create_readcondition(reader1, DDS_ANY_STATE);
+          memset(samples1, 0x0, sizeof(samples1));
+          dds_waitset_attach(ws1, rcond1, rcond1);
+
+          int j;
+          int cnt = 0;
+          bool match = true;
+          while ( match && cnt < sample_cnt ) {
+            result = dds_waitset_wait(ws1, NULL, 0, DDS_SECS(10));
+            CU_ASSERT_FATAL(result > 0);
+            result = dds_take(reader1, (void**)samples1, sinfo, sample_cnt, sample_cnt);
+            CU_ASSERT_FATAL(result > 0);
+            for (int i = 0; i < result; ++i) {
+              j = cnt + i;
+              if (
+                samples1[i]->a != j ||
+                samples1[i]->b != j + 1 ||
+                samples1[i]->c != j + 2
+              ) {
+                match = false;
+              }
+            }
+            cnt += result;
+          }
+          CU_ASSERT_FATAL(match);
+          CU_ASSERT_FATAL(cnt == sample_cnt);
+        }
+        {
+          PsmxType1* samples2[sample_cnt];
+          dds_sample_info_t sinfo[sample_cnt];
+          const dds_entity_t ws2 = dds_create_waitset(participant);
+          dds_entity_t rcond2 = dds_create_readcondition(reader2, DDS_ANY_STATE);
+          memset(samples2, 0x0, sizeof(samples2));
+          dds_waitset_attach(ws2, rcond2, rcond2);
+
+          int j;
+          int cnt = 0;
+          bool match = true;
+          while ( match && cnt < sample_cnt ) {
+            result = dds_waitset_wait(ws2, NULL, 0, DDS_SECS(10));
+            CU_ASSERT_FATAL(result > 0);
+            result = dds_take(reader2, (void**)samples2, sinfo, sample_cnt, sample_cnt);
+            CU_ASSERT_FATAL(result > 0);
+            for (int i = 0; i < result; ++i) {
+              j = cnt + i;
+              if (
+                samples2[i]->xy.x != j ||
+                samples2[i]->xy.y != (uint8_t)j + 1 ||
+                samples2[i]->z != (uint8_t)j + 2
+              ) {
+                match = false;
+              }
+            }
+            cnt += result;
+          }
+          CU_ASSERT_FATAL(match);
+          CU_ASSERT_FATAL(cnt == sample_cnt);
+        }
+      }
+      dds_delete(domain);
+    }
+    {
+      // Test case using a config without specifying a locator.
```

Comment:
Same as above: use the helper function and avoid assumptions on plugin internals.